### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by edmorley**
The initial implementation of a native (non-[shimmed](https://github.com/heroku/cnb-shim)) [libcnb.rs](https://github.com/heroku/libcnb.rs)-based Python CNB.

See:
- [Project overview](https://salesforce.quip.com/6xkhArt5bqCS) (internal)
- [GUS epics/work report](https://gus.lightning.force.com/lightning/r/Report/00OAH000000JqVj2AK/view) (internal)
- Work that is out of scope for the initial version: https://github.com/heroku/buildpacks-python/issues
- Upstream Cloud Native Buildpacks information:
  -  https://buildpacks.io/docs/
  - https://github.com/buildpacks/spec/blob/main/buildpack.md
- [libcnb.rs documentation](https://docs.rs/libcnb/latest/libcnb/)
- Prior art:
  - https://github.com/heroku/heroku-buildpack-python (the classic Heroku Python buildpack)
  - Other `libcnb.rs`-based Heroku CNBs
    - https://github.com/heroku/buildpacks-go
    - https://github.com/heroku/buildpacks-jvm
    - https://github.com/heroku/buildpacks-nodejs
    - https://github.com/heroku/procfile-cnb

### CODE REVIEW

File: src/python_version.rs
Comment (Malax) [inline src/python_version.rs:32]: Nit, non-blocking: Personally, I would move the download URL templating out of `PythonVersion` to increase separation of concerns.

File: src/packaging_tool_versions.rs
Comment (Malax) [inline src/packaging_tool_versions.rs:9]: Question: Are there better ways of representing these versions than `String`? For example, if they're SemVer we could leverage the `semver` crate. However, if all we do in this buildpack is comparing them verbatim `String` seems fine.

**Review by Malax — APPROVED** [review]
The README could use some work, but this can easily be added later. That said, I think this is the new gold-standard for a Heroku languages-team CNB! I love the code, documentation and tests and wish my buildpack would look like this. It was a pleasure reviewing this! Awesome work dude! ✋🏻 

I left a couple of nitpicky comments, but none are blocking at all.

File: tests/integration/pip.rs
Comment (edmorley) [reply tests/integration/pip.rs:2]: They are used inside the `formatdoc!` blocks (Clippy should report anything that's unused too 🤞)

File: tests/integration/python_version.rs
Comment (edmorley) [reply tests/integration/python_version.rs:1]: Ah yes this stray comment was from when all of the integration tests were in a single file, and comments were used to categorise the tests. I'll remove.

File: tests/integration/python_version.rs
Comment (edmorley) [reply tests/integration/python_version.rs:4]: It's used in the `formatdoc!` blocks

File: src/packaging_tool_versions.rs
Comment (edmorley) [reply src/packaging_tool_versions.rs:9]: These versions don't use semver (for example, Pip doesn't use `X.Y.0` but `X.Y` for the first release of a new minor). They are only used as strings, and also the version handling here will change as part of [Reduce toil required to keep pip/setuptools/wheel up to date](https://gus.lightning.force.com/lightning/r/ADM_Epic__c/a3QEE000000WdUT2A0/view) (which will switch to pulling the values from requirements files, so Dependabot can keep them up to date).

File: src/python_version.rs
Comment (edmorley) [reply src/python_version.rs:32]: Yeah I wasn't 100% happy with it being here (previously the URL was inlined in the layer itself) - although this is all going to change once the versions manifest file is added as part of [Automatic Python patch version updates](https://gus.lightning.force.com/lightning/r/ADM_Epic__c/a3QEE000000V4tl2AC/related/Work__r/view).

File: tests/integration/python_version.rs
Comment (edmorley) [reply tests/integration/python_version.rs]: Removed in b189360053295c675cf23eb3b8476d6162d208a4

File: src/packaging_tool_versions.rs
Comment (merwok) [inline src/packaging_tool_versions.rs:6]: FWIW, it you needed to work with the versions eventually, someone implemented the python packaging version spec: https://crates.io/crates/pep440

File: src/main.rs
Comment (schneems) [inline src/main.rs:100]: You're passing in the current environment. I think you also need to apply `context.platform` environment (assuming that users will want to do things like use env vars to do things, like setting private resource creds).

File: src/layers/python.rs
Comment (schneems) [inline src/layers/python.rs:33]: It's a bit annoying working with structs that don't own their data. Cloning is minimal overhead in this context and I'm mindful that people who primarily dev in python (and not in Rust) may try to submit a PR and that might be a stumbling block. (For an analogy: Perhaps you remember how annoying working on libcnb internals with the generic in `libcnb::Error<E>` was when first getting started). 

I think that using a reference here might be the more "rust-y" way to do things, (clone is bad, burn with fire 🔥). But in our scripting context, I think (some) dev speed takes precedence. Shorter: I don't think the gain of no clone is worth having to put up with lifetime annotations for the rest of the file.

File: src/layers/python.rs
Comment (schneems) [inline src/layers/python.rs:32]: Thank you for the docs 🙏

File: src/layers/python.rs
Comment (schneems) [inline src/layers/python.rs:224]: I really like this approach of building up all the reasons it's changed (rather than limiting to only one). I would love to come up with some way to make this easier/automagical for users via some trait or macro or something. Maybe manuel (or you) might have some ideas on interface and implementation.

It seems most of us have converged on similar patterns, it would be nice to make the common/default case require little to no code.

File: src/layers/python.rs
Comment (schneems) [inline src/layers/python.rs:219]: I love the destructuring pattern to ensure all values are used. Style wise I don't think you need it twice though. You're guaranteed to have the same fields in each. You could:

```rust
    let PythonLayerMetadata {
        stack,
        python_version,
        packaging_tool_versions:
            PackagingToolVersions {
                pip_version,
                setuptools_version,
                wheel_version,
            },
    } = new_metadata;

    if stack != cached.stack {
        reasons.push(format!(
```

It would save some lines.

File: src/packaging_tool_versions.rs
Comment (schneems) [inline src/packaging_tool_versions.rs:7]: `Debug` all structs by default please.
```suggestion
#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
```

File: src/layers/pip_cache.rs
Comment (schneems) [inline src/layers/pip_cache.rs:14]: `Debug` all structs by default please.

File: src/layers/pip_cache.rs
Comment (schneems) [inline src/layers/pip_cache.rs:75]: `Debug` all structs by default please.

```suggestion
#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
```

File: src/layers/pip_dependencies.rs
Comment (schneems) [inline src/layers/pip_dependencies.rs:15]: `Debug` all structs by default please.

File: src/layers/python.rs
Comment (schneems) [reply src/layers/python.rs:33]: Also Debug all structs by default please.

File: src/layers/python.rs
Comment (schneems) [inline src/layers/python.rs:183]: `Debug` all structs by default please.

```suggestion
#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
```

File: src/main.rs
Comment (schneems) [inline src/main.rs:35]: Debatable but for consistency sake: `Debug` all structs by default please.

File: src/layers/pip_dependencies.rs
Comment (schneems) [reply src/layers/pip_dependencies.rs:15]: I bet we could write a custom lint for this.

**Review by schneems — DISMISSED** [review]
I'm still working on finishing my review, but wanted to leave my pending comments.

File: src/utils.rs
Comment (edmorley) [reply src/utils.rs:31]: > Errors from std::io::Error do not include the target file or what operation failed. I've been suggesting use of the fs_err crate https://docs.rs/fs-err/latest/fs_err/index.html

Yeah, I wish the stdlib included paths too, I am subscribed to (and have a thumbs up on):
https://github.com/rust-lang/rfcs/issues/2885
https://github.com/rust-lang/rust/pull/86826

...and had seen your usage of the `fs_err` crate in the Ruby CNB and checked it out, since it looks interesting.

> though camino is another popular alternative 
> ...
> It looks like camino has this method implemented https://docs.rs/camino/1.1.3/camino/struct.Utf8Path.html#method.try_exists

I don't see how Camino solves the filename problem? It uses the stdlib `io::Error`?
https://docs.rs/camino/1.1.3/src/camino/lib.rs.html#1335

If we wanted to switch wholesale to camino, then I think this should start within libcnb.rs first.

> I think this gives a sub-optimal error message

I considered this when implementing. I tried very hard to break an app/build in order to write an integration test that tests the detection IO error message. However, I couldn't find a way to do so, since all the obvious approaches (such as `chmod`ing a file locally to remove permissions) caused Pack / `lifecycle` itself to fail, and never got as far as the buildpack. 

As such, I'm confident that this detect IO error will always only ever be due to a platform issue (eg the whole EBS mount is messed up), rather than an application source issue (eg wrong permissions on say `requirements.txt`). Therefore, having the filename be listed here doesn't add much value.

If you can demonstrate a single integration testcase that proves this is not the case, then I will happily reconsider :-)

> If they have one bad and one good file, we still fail the build as this try ? will fail eagerly. 

Yup, this is intentional. We shouldn't silently ignore errors here.

> I think we should actually not care about if the file is valid or not here at all, but always return a successful detection, then deliver a better error message in the build phase.

Pack already shows the full detect error if the build fails, so we don't need to defer to the build phase for this.

**Review by edmorley — COMMENTED** [review]
@schneems Thank you for taking the time to review! :-)

File: src/packaging_tool_versions.rs
Comment (edmorley) [reply src/packaging_tool_versions.rs]: @merwok Thank you - good spot! Fixed in 7dce0d59152e12228735f45784b2fa5c6744a99e

File: src/packaging_tool_versions.rs
Comment (edmorley) [reply src/packaging_tool_versions.rs:6]: @merwok Yeah I'd spotted that crate while looking for another use-case - great that it exists! (Albeit sadly it only parses single versions and not ranges, so it can't be used for things like parsing python version ranges from the PEP440 style versions in a Poetry lockfile.)

File: src/utils.rs
Comment (schneems) [reply src/utils.rs:31]: > I don't see how Camino solves the filename problem? It uses the stdlib io::Error?

Sorry. I was parroting some community advice. I heard about both at the same time and was under the impression that camino gave better error messages, but either I misunderstood, or it's not applied to all contexts. I would need to actually exercise this code path to make sure they're not providing that functionality in some sneaky way i'm just not seeing.

> If you can demonstrate a single integration testcase that proves this is not the case, then I will happily reconsider :-)

Did you only try changing permissions or also symlinks? Here's an example of a bad symlink https://github.com/schneems/which_problem/blob/65bda424cf62bc04aa23f192e971a80e4db0ae83/which_problem/src/lib.rs#L148-L153

>  Therefore, having the filename be listed here doesn't add much value.

I think it's pretty core to the debugging process. You're operating under the assumption that this can only fail of something catastrophic happened and ALL files are failing. I'm not as confident. Even if it's not user trigerable, the first thing I would want to know if I was debugging this as a buildpack author or pack maintainer is: What file failed.

Comment (schneems) [reply src/utils.rs:31]: Scratch that. I understand now. Sorry that last response was missing context on my part. The `try_exist` fails based on permissions issues, only for parent directories, not for self. It does not Err on symlinks. 

So yes, i'm confident that this would be a CNB bug and not a user issue (now). I also tried out camino and confirm it doesn't help at all with an error message in this case:

```rust
fn main() {}

#[cfg(test)]
mod tests {
    use std::ffi::OsString;

    #[test]
    fn lol() {
        let tmp_dir = tempfile::tempdir().unwrap();
        let dir = tmp_dir.path();
        let file = dir.join("lol");

        std::fs::write(&file, "lol");

        use std::os::unix::fs::PermissionsExt;
        let metadata = dir.metadata().unwrap();
        let mut perms = metadata.permissions();
        perms.set_mode(0o000);

        std::fs::set_permissions(&dir, perms).unwrap();

        camino::Utf8Path::new(file.to_str().unwrap())
            .try_exists()
            .unwrap();
    }
}
```

Gives us:

```
thread 'tests::lol' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', src/main.rs:24:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The error message raised here is:

```
        BuildpackError::DetectIo(io_error) => log_io_error(
            "Unable to complete buildpack detection",
            "determining if the Python buildpack should be run for this application",
            &io_error,
        ),
```

Which will show up as:

```
Unable to complete buildpack detection

An unexpected error occurred whilst determining if the Python buildpack should be run for this application.
            
Details: I/O Error: Permission denied
```

To an end user it would be unclear what permissions are denied (parent directory does not have read permissions) and that it's an error with pack/CNB/libcnb and not with their application.

File: src/package_manager.rs
Comment (schneems) [reply src/package_manager.rs:18]: Sounds good.  We can revisit then.

File: src/layers/python.rs
Comment (schneems) [reply src/layers/python.rs:33]: Bikeshed should be blue! It's not a blocker. Feel free to resolve.

It's a bit of an open question how similar should buildpack internal/style be? Should we update all buildpack layers use non-owned data?

File: src/layers/python.rs
Comment (schneems) [reply src/layers/python.rs:224]: > That said, I quite like having human readable explanations for the attributes, at which point any abstraction here would have to take that into account too (and so any abstraction may be unwieldy and not worth doing).

My philosophy is: We won't know until we try. Implementation is iteration. 

If we could metaprogram and dynamically get the field names we could use a convention of a default explanation based on the field name. Something like:

```
Detected changes, clearing the cache:
- Stack heroku-18 changed to heroku-22
- Python version 2.7 changed to 3.0
```

I think we could maybe codegen something via a macro which would (should?) have access to field names.

Of course we could still all write our own bespoke comparisons and messages, but it would be nice to have the option of a really solid/easy default, even if it's not perfect.

File: src/layers/python.rs
Comment (schneems) [reply src/layers/python.rs:219]: Ahh, true. Now that you mention it. I ran into the same issue, but decided to use individual let statements inside of my branch.

File: src/layers/pip_cache.rs
Comment (schneems) [reply src/layers/pip_cache.rs:14]: Not having it makes println and dbg based debugging harder, and (in my opinion) surprising. If I'm in your codebase and need it to debug something, and add it to one struct, then it has to cascade to all other structs that it holds that also don't implement debug. Then once I finally fix the bug I was chasing, I'm left with two options:

- Submit a messy PR with a bunch of Debug additions and my fix.
- Carefully remove all the Debug additions and hope that I don't have to add them back in after someone realizes there's an edge case that needs to be handled.

With Rust print/puts debugging is less needed, but it's not an absurd or exotic use case.

File: src/main.rs
Comment (schneems) [reply src/main.rs:100]: Sounds intentional 👍 

Wording wise "due to the use of `clear-env` in `buildpack.toml`. " is not terribly informative for me. Looking up the docs:

```
When clear-env in buildpack.toml is set to true for a given buildpack, the lifecycle MUST NOT set user-provided environment variables in the environment of /bin/detect or /bin/build.
```

I think that `clear-env` info is an implementation detail of the exclusion. I would say the more important reason is WHY they're being excluded. Maybe something like:

```
All user-provided env vars are intentionally excluded to prevent users from accidentally setting an unexpected configuration value. This behavior matches the classic buildpack.
```

File: src/utils.rs
Comment (joshwlewis) [inline src/utils.rs:56]: While you are adding TODOs - maybe one to verify the download against a known checksum? This might be more of an epic though, considering the existing comments around adding and using an inventory/manifiest.

**Review by joshwlewis — APPROVED** [review]
Looks awesome! :shipit:

File: src/layers/python.rs
Comment (edmorley) [reply src/layers/python.rs:33]: For some things there is a definite advantage in standardising, particularly when there is either a clear best practice (eg: "use rustfmt on our codebases") or the decision matters from a product point of view (eg: consistent log output styles between buildpacks).

However, I think there are always going to be things that fall outside of the scope of linters/formatters and just come down to whatever implementation details happen to have been chosen - e.g. use of an if statement vs a match statement, or in this case whether a struct owns its data or not. For these I'm not sure I see the benefit on standardising - and also maybe there are more important things to discuss instead of them, during review?

File: src/layers/pip_cache.rs
Comment (edmorley) [reply src/layers/pip_cache.rs:14]: I agree that is one case where having `Debug` on internal things can be useful.

However:
- I rarely use println debugging (instead unit test, or the debugger)
- It very easy to exclude print statements or things added when developing using partial Git stating (I use GitKraken, but there are lots of ways to do this)
- The structs in this buildpack aren't super complicated/nested, so the "I need to cascade Debug everywhere" issue doesn't apply here

If this ever comes up in reality as being a problem for this buildpack, then I'm happy to reconsider :-)

File: src/main.rs
Comment (edmorley) [reply src/main.rs:102]: 192230568b35c7ac32e20677e4d11838dc8b0659

File: .github/workflows/ci.yml
Comment (edmorley) [reply .github/workflows/ci.yml:16]: @joshwlewis `libcnb-test` performs the same packaging process that `cargo libcnb package` does (the only difference is that for actual releases we use `--release` for `cargo libcnb package`). As such, for buildpacks that are not meta-buildpacks I don't think there is any benefit to running `cargo libcnb package` directly. Good spot though! :-)

File: src/utils.rs
Comment (edmorley) [reply src/utils.rs:56]: Yeah I have this epic for that:
[Python CNB: Validate the SHA256 checksum of Python binaries](https://gus.lightning.force.com/lightning/r/ADM_Epic__c/a3QEE000000V4vN2AS/view)

File: src/utils.rs
Comment (edmorley) [reply src/utils.rs:31]: In the buildpack I'm using the phase "unexpected error" to cover things that are "not normal, potentially our issue but it's hard to say more than that". I have an existing TODO to explore this further:
https://github.com/heroku/buildpacks-python/blob/192230568b35c7ac32e20677e4d11838dc8b0659/src/errors.rs#L280

For now, what's there is more than enough (and a better implementation already than several of the other CNBs), and I don't see this as a blocker, so given there is a TODO I'm going to resolve this for now.

Thank you for raising some good points however - I'll bear them in mind for the future :-)

**Pull request merged**
